testtools.py

```diff
@@ -0,0 +1,254 @@
+
+    def destroy(self) -> None:
+        """
+        Destroy all nodes in the cluster.
+        """
+        self._make(variables={}, target='clean')
+
+    def _nodes(self, container_base_name: str, num_nodes: int) -> Set[_Node]:
+        """
+        Args:
+            container_base_name: The start of the container names.
+            num_nodes: The number of nodes.
+
+        Returns: ``_Node``s corresponding to containers with names starting
+            with ``container_base_name``.
+        """
+        client = Client()
+        nodes = set([])  # type: Set[_Node]
+
+        while len(nodes) < num_nodes:
+            container_name = '{container_base_name}{number}'.format(
+                container_base_name=container_base_name,
+                number=len(nodes) + 1,
+            )
+            details = client.inspect_container(container=container_name)
+            ip_address = details['NetworkSettings']['IPAddress']
+            node = _Node(
+                ip_address=ip_address,
+                ssh_key_path=self._path / 'include' / 'ssh' / 'id_rsa',
+            )
+            nodes.add(node)
+
+        return nodes
+
+    @property
+    def masters(self) -> Set[_Node]:
+        """
+        Return all DC/OS master ``_Node``s.
+        """
+        return self._nodes(
+            container_base_name='dcos-docker-master',
+            num_nodes=self._masters,
+        )
+
+
+class Cluster(ContextDecorator):
+    """
+    A record of a DC/OS Cluster.
+
+    This is intended to be used as context manager.
+    """
+
+    def __init__(
+        self,
+        extra_config: Dict,
+        masters: int=1,
+        agents: int=0,
+        public_agents: int=0,
+    ) -> None:
+        """
+        Args:
+            extra_config: This dictionary can contain extra installation
+                configuration variables to add to base configurations.
+            masters: The number of master nodes to create.
+            agents: The number of master nodes to create.
+            public_agents: The number of master nodes to create.
+        """
+
+        # See README.md for information on the required configuration.
+        with open('configuration.yaml') as configuration:
+            tests_config = yaml.load(configuration)
+
+        generate_config_path = None
+
+        if tests_config['dcos_generate_config_path'] is not None:
+            generate_config_path = Path(
+                tests_config['dcos_generate_config_path']
+            )
+
+        self._backend = _DCOS_Docker(
+            masters=masters,
+            agents=agents,
+            public_agents=public_agents,
+            extra_config=extra_config,
+            generate_config_url=tests_config['dcos_generate_config_url'],
+            generate_config_path=generate_config_path,
+            dcos_docker_path=Path(tests_config['dcos_docker_path']),
+        )
+        self._backend.postflight()
+
+    def __enter__(self) -> 'Cluster':
```

Comment:
Because this is in the `Cluster` class, `Cluster` is not defined at the time that this is defined. This seems to be the best way to define this type. See _numerous_ discussions online e.g. https://github.com/python/typing/issues/34.